### ISSUE TITLE
Run Program should support mail body

### ISSUE THREAD

**Issue by WebFreak001** [RealRaven2000/FiltaQuilla#41]
For the Run Program action there should be an option to dump the mail body into the program stdin.

Use-case: I want to parse a value inside the mail and by that operate on some local files or issue HTTP calls and do whatever I want in my existing application built for this.

To allow attachments and even more handling it might be useful to have an advanced mode to directly dump the whole raw email (headers + content + multipart encoded attachments) into the program.

**Comment by RealRaven2000** [comment]
> To allow attachments and even more handling it might be useful to have an advanced mode to directly dump the whole raw email (headers + content + multipart encoded attachments) into the program.

How would you go about transmitting the data? There may be limits for command lines, so does it make sense to automatically export the mail to a HTML file and then passing the full file name / path to the program.

**Comment by WebFreak001** [comment]
> dump the mail body into the program stdin

I think the best way to do this is by piping all the content into stdin, it would also not break any previous programs like this. This also pretty much reflects the unix way of doing things and can easily automate things inside scripts like this with common utilities such as grep.

I looked into it but with the current XPCOM calls this will not be possible. Instead it will need to bundle a native application (shell script + batch script) and prompt the user to install it when enabling the run program checkbox. Then it can use the native messaging API. This will also be future-proof in case the currently used process API gets removed.

**Comment by RealRaven2000** [comment]
> > dump the mail body into the program stdin
> 
> I think the best way to do this is by piping all the content into stdin, it would also not break any previous programs like this. This also pretty much reflects the unix way of doing things and can easily automate things inside scripts like this with common utilities such as grep.

I currently don't have the bandwidth to do this, but if you are a programmer you're free to contribute a patch. I am not a great fan of bundling external / OS-specific software with Add-ons because I cannot test this on Unix or Mac, and it also causes problems when reviewing on ATN, so using XPCOM is my personal preference. I am aware that this method will be deprecated and need to be replaced with APIs. 

I am actively lobbying for file access support for Add-ons, I think there is actually a feature planned by Mozilla that gives specific permissions for file access on a per-folder basis. That's why I personally would use a file interface.

**Comment by RealRaven2000** [comment]
I wonder if we can save the email as eml format asynchronously and then spawn the call to the external program.... My main problem is whether it is even safe to save the email during filter running as it is expected to be a synchronous operating (as designed in Thunderbird) and it might hang the whole filtering process to do this.

**Comment by RealRaven2000** [comment]
I agree we should have @BODY@ as a parameter for Run Program, not sure how to achieve this completely, hopefully somebody is still interested now that we are already in Tb 115 era
